src/main/scala/com/microsoft/hyperspace/Hyperspace.scala

```diff
@@ -166,18 +168,20 @@ class Hyperspace(spark: SparkSession) {
 }
 
 object Hyperspace {
-  private lazy val context = new ThreadLocal[HyperspaceContext]
+  // Maintain HyperspaceContext for each Spark session, since Hyperspace depends on
+  // the session's properties such as configs, etc.
+  private[hyperspace] lazy val contexts =
+    mutable.WeakHashMap[SparkSession, ThreadLocal[HyperspaceContext]]()
```

Comment:
@imback82 is this ThreadLocal required? if there are several threads holding the same spark session, then yes?
I'm bit confused about previous use - `ThreadLocal` in `object`. If that is working properly, is it ok to use the previous ver. when each thread has its own spark session?

+ I tested weakPtr behavior (with System.gc + waiting), but it seems Spark (sharedState?) holds the reference of each session even after I removed the reference in test code; `WeakHashMap` might be ineffective.

Reply:
So, `SparkSession` created from `newSession` can never be GC-ed? If that's the case, we don't need to use `WeakHashMap`.

Reply:
Yep I was waiting more than 30 mins, but no gc-ed. I guess it's because of statistics or history info. 
Then I'll replace it with ConcurrentHashMap

Reply:
sgtm